### PR TITLE
Private chains

### PR DESCRIPTION
**Pull request by ajunge**
Support for testnets and private chains

### CODE REVIEW

**Comment by awoie** [comment]
@ajunge should we make use of the EIP155 registry for the RPC endpoints? further, would it make sense to have a "connector" object that handles the connection with the specific network? we could preregister "connectors" for all networks were we deployed our ERC1056. for private chains, a new "connector" could then be registered.

File: src/networks.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        "rpcUrl": "https://kovan.infura.io/ethr-did",
+        "registry": "0xdca7ef03e98e0dc2b855be647c39abe984fcf21b"
+    },
+    {
+        "name": "rsk",
+        "rpcUrl": "https://public-node.rsk.co",
+        "registry": "0xdca7ef03e98e0dc2b855be647c39abe984fcf21b"
+    },
+    {
+        "name": "rsk:testnet",
```
[the remark this answers is not in the document]
Comment (ajunge) [reply src/networks.json]: Alastria does not have public endpoint where we can point the resolver.

File: src/__tests__/networks-test.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                authentication: [
+                    {
+                        type: 'Secp256k1SignatureAuthentication2018',
+                        publicKey: `${did}#owner`,
+                    },
+                ],
+            })
+        })
+
+        it.skip('resolves on alastria:telsius', () => {
```
[the remark this answers is not in the document]
Comment (ajunge) [reply src/__tests__/networks-test.js]: I need to delete those test. Neither alastria or rsk have public rpc endpoints we can use to resolve. The tests where there to try with private endpoint I have

**Comment by mi-xu** [comment]
Changes made by this PR:
- added documentation explaining how to configure ethr-did-resolver to use an alternate Ethereum network
- added ability to specify ethr-did-registry contract address as part of the network configuration

**Comment by ajunge** [comment]
> * If the RSK networks are not working at the moment let's remove them from the documentation, configs, and tests so that we are not exposing things that will break
> * If there is no public Alastria network config let's remove those from the documentation, configs, and tests as well
> * Can we configure the linter to respect the original 2 space styling?  Changing it to 4 seems unnecessary and obfuscates the diff view of any changes that were made to the actual code.

done

File: src/register.js [hunk 3]
```diff
@@ -90,7 +91,7 @@ export function wrapDidDocument(did, owner, history) {
           const type = attrTypes[match[4]] || match[4]
           const encoding = match[6]
           switch (section) {
-            case 'pub':
+            case 'pub': {
```
Comment (mi-xu) [inline src/register.js]: Are these curly braces around the case body necessary?  If not can we not add them them extraneously and maintain consistency with the rest of the code?

File: src/register.js [hunk 4]
```diff
@@ -141,10 +143,10 @@ export function wrapDidDocument(did, owner, history) {
     } else {
       if (
         delegateCount > 0 &&
-        (event._eventName === 'DIDDelegateChanged' ||
-          (event._eventName === 'DIDAttributeChanged' &&
-            bytes32toString(event.name).match(/^did\/pub\//))) &&
-        validTo.lt(now)
+    (event._eventName === 'DIDDelegateChanged' ||
```
Comment (mi-xu) [inline src/register.js]: Statements that span multiple lines being indented less than the starting line is confusing and inconsistent with the original code style.  Can we revert these changes since they are unnecessary and reduce readability?  This happens in a few other places as well (line 19-22, line 62).

File: src/register.js [hunk 5]
```diff
@@ -176,27 +178,49 @@ function configureProvider(conf = {}) {
...
+}
+
+export default function register(conf = {}) {
+  
   const logDecoder = abi.logDecoder(DidRegistryContract, false)
 
-  const lastChanged = async identity => {
-    const result = await didReg.changed(identity)
+  const networks={
+    'mainnet': configureNetwork(conf),
```
Comment (mi-xu) [inline src/register.js]: This is rather confusing to configure mainnet like this using the `conf` object and then also configuring multiple networks using `conf.networks` later.  Why not include the mainnet config as part of the `networks.json` instead of maintaining a unique edge case?

File: src/register.js [hunk 6]
```diff
@@ -176,27 +178,49 @@ function configureProvider(conf = {}) {
...
   const provider = configureProvider(conf)
   const eth = new Eth(provider)
   const registryAddress = conf.registry || REGISTRY
   const DidReg = new EthContract(eth)(DidRegistryContract)
   const didReg = DidReg.at(registryAddress)
+  
+  return { eth, registryAddress, didReg}
+}
+
+function confNetworks(networksConf=[]){
```
Comment (mi-xu) [inline src/register.js]: Naming is inconsistent by shortening to "confs" here compared to "configureNetwork."

File: README.md [hunk 7]
```diff
@@ -187,12 +189,49 @@ The resolver presents a simple `resolver()` function that returns a ES6 Promise
...
+| `networks[0].registry` | Address of the registry on the private network                               |
+
+Either provider, web3 or rpcUrl is needed.
+
+```js
+import resolve from 'did-resolver'
+import register from 'ethr-did-resolver'
+
+const registyAddress='0x....' //Registry address on privChain
+const jsonRpcEndpoint='https://priv.chain/rpc' //JSON-RPC endpoint for privChain
```
Comment (mirceanis) [inline README.md]: This configuration does not seem to explicitly support endpoints that require authentication.
Is there any guidance about how that should be handled?

Comment (ajunge) [reply README.md]: No support for authenticated endpoints in this PR

File: src/register.js (hunk 3 above)
Comment (ajunge) [reply src/register.js]: I didn't touch this part. Is from the previous code already. Should I fix it anyways?

File: src/register.js (hunk 4 above)
Comment (ajunge) [reply src/register.js]: I didn't touch this part. Is from the previous code already. Should I fix it anyways?

File: src/register.js (hunk 3 above)
Comment (mi-xu) [reply src/register.js]: It is showing as a change in the diff

File: src/register.js (hunk 4 above)
Comment (mi-xu) [reply src/register.js]: It is showing as a change in the diff

File: src/register.js (hunk 6 above)
Comment (ajunge) [reply src/register.js]: What should be the proper name?

Comment (mirceanis) [reply src/register.js]: how about `configureNetworks` ?

File: src/networks.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
+[
+  {
+    "name": "ropsten",
+    "rpcUrl": "https://ropsten.infura.io/ethr-did",
```
Comment (mirceanis) [inline src/networks.json]: these are old infura URLs and don't use the new project based address.

I don't think we should be publishing these defaults.. or at least not as is.

File: src/ethr-did-resolver.js [hunk 9]
```diff
@@ -216,9 +237,18 @@ function getResolver (conf = {}) {
...
+      throw new Error(`Not a valid ethr DID: ${did}`)
+
+    const id=fullId[2];
+    const network = (!fullId[1]) ? 'mainnet' : fullId[1].slice(0,-1);
+
+    if(!networks[network])      
+      throw new Error(`No conf for network: ${network}`)
+
+    const { owner, history } = await changeLog(id)
     return wrapDidDocument(did, owner, history)
```
Comment (mirceanis) [inline src/ethr-did-resolver.js]: I believe the DID document for `did:ethr:0xaddress` and `did:ethr:mainnet:0xaddress` should be identical.
If this is not the case then the kotlin implementation should also be adapted to fit
